### ISSUE TITLE
Include TestScheduler.flush() in toBe function

### ISSUE THREAD

**Issue by arichiardi** [ReactiveX/rxjs#2719]
Hello folks!

I am trying to wrap my head around marble testing and so far, with bumps here and there, I was able to make it works.
The only thing I was wondering is why I need to call `TestSchedules.flush()` explicitely in the tests for it to work. Maybe I am missing something, in which case I am more than willing to contribute back to the docs but now in my test helpers I hacked this together:

```typescript
export function expectObservable(observable: rx.Observable<any>, unsubscriptionMarbles: string = null): observableToBeObject {
  const origToBe = rxTestScheduler.expectObservable(observable, unsubscriptionMarbles);
  // we don't want to call flush() every time, I wonder why this is not in RxJS already
  return {
    toBe: (marbles, values?, errorValue?) => {
      origToBe.toBe(marbles, values, errorValue)
      rxTestScheduler.flush();
    }
  }
}
```

Any advice, suggestion,tips on this? Is it a bad idea to include the `flush()` in the `toBe` like above?

Thanks!

**Comment by kwonoj** [comment]
Cause testscheduler's flush mutate states of hot/cold observable. It is not recommended cases though, if assertion is written in this way
```
hot = ...
expectObservable(hot).toBe()
another = ...
expectObservable(hot).toBe()
```
now implicit flush will cause some unexpected behavior as flush intended to be happen only when assertion begins, all assertions are already planned at the moment. If all of test case is written to satisfy those then it won't be big matter cause flushing multiple times without setting another assertions in between. So instead, our test case overrides execution context itself (`it` function) and ensure flush occurs at the end of assertion always.

We have long term plan to rewrite test scheduler with more fine grained interface and thinking about time advancing link v4 interfaces, then it'll be much more important to decide when to flush scheduler.

In short, it's ok in your test case but will not be in core.

**Comment by arichiardi** [comment]
@kwonoj thanks for the explanation!
So do I understand right by saying that I should always `flush()` at the end of the test case?

**Comment by kwonoj** [comment]
To avoid unexpected assertion side effect, yes. Otherwise when it flushes it emits all hot/cold observable scheduled in prior.
